pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -2052,6 +2052,69 @@ def usage(self) -> _usage.Usage:
         """Get usage statistics for the run so far, including token usage, model requests, and so on."""
         return self._graph_run.state.usage
 
+    async def validate_structured_output(
+        self, response: _messages.ModelResponse, *, tool_name: str | None = None, allow_partial: bool = False
+    ) -> FinalResult[OutputDataT] | None:
+        """Validate a structured result message.
+
+        This is analogous to `pydantic_ai.result.StreamedRunResult.validate_structured_output`, but can be used
+        when iterating over an agent run even when you aren't on the final response.
+
+        Args:
+            response: The structured result message.
+            tool_name: If provided, this should be the name of the tool that will produce the output.
+                This is only included so that you can skip the tool-lookup step if you already know which tool
+                produced the output, which may be the case if calling this method in a loop while streaming.
+                (You can get the `tool_name` from the return value of this method when it is not `None`.)
```

Comment:
Not sure if we really want to add the complexity associated with allowing the end user to specify the tool, or just always look up the tool here. I'd be okay either way, dropping this should be trivial though if we want to (just delete the argument and the first `if`-statement referencing it below).